lib/Sema/CSSolver.cpp

```diff
@@ -713,7 +713,11 @@ bool ConstraintSystem::tryTypeVariableBindings(
           newBindings.push_back({subtype, binding.Kind, binding.BindingSource});
       }
 
-      if (binding.Kind == AllowedBindingKind::Subtypes) {
+      // Allow solving for T even for a binding kind where that's invalid
+      // if fixes are allowed, because that gives us the opportunity to
+      // match T? values to the T binding by adding an unwrap fix.
+      if (binding.Kind == AllowedBindingKind::Subtypes ||
+          shouldAttemptFixes()) {
```

Comment:
At first glance I'm not sure this has to be guarded by `shouldAttemptFixes()`, and it seems like it might even be misleading to do so because we're not recording a fix here.

It also seems like if this is legal (and again at first glance I think it is), this should be part of a separate PR.

Reply:
I'd be fine with pulling out the last two commits here into a separate PR and merging the rest to get it in, if you are ok with that.

There is no functional change for the `shouldAttemptFixes()` check here because this attempted binding will just be discarded if `solveRec()` fails due to lack of fixes. But I don't think that it can succeed in the non-subtypes case without fixes so it is saving time.

Reply:
I think this is kind of chicken-and-egg problem here, we have to attempt to produce a fix and we can't record it upfront either...